setuptools/discovery.py

```diff
@@ -0,0 +1,396 @@
+
+
+# We have to be extra careful in the case of flat layout to not include files
+# and directories not meant for distribution (e.g. tool-related)
+
+
+class FlatLayoutPackageFinder(PEP420PackageFinder):
+    _EXCLUDE = (
+        "bin",
+        "doc",
+        "docs",
+        "documentation",
+        "test",
+        "tests",
+        "example",
+        "examples",
+        "scripts",
+        "tools",
+        # ---- Task runners / Build tools ----
+        "tasks",  # invoke
+        "fabfile",  # fabric
+        "site_scons",  # SCons
+        # ---- Hidden directories/Private packages ----
+        "[._]*",
+    )
+
+    DEFAULT_EXCLUDE = tuple(chain_iter((p, f"{p}.*") for p in _EXCLUDE))
+    """Reserved package names"""
+
+    _looks_like_package = staticmethod(_valid_name)
+
+
+class FlatLayoutModuleFinder(ModuleFinder):
+    DEFAULT_EXCLUDE = (
+        "setup",
+        "conftest",
+        "test",
+        "tests",
+        "example",
+        "examples",
+        "build",
+        # ---- Task runners ----
+        "noxfile",
+        "pavement",
+        "dodo",
+        "tasks",
+        "fabfile",
+        # ---- Other tools ----
+        "[Ss][Cc]onstruct",  # SCons
+        "conanfile",  # Connan: C/C++ build tool
+        "manage",  # Django
+        # ---- Hidden files/Private modules ----
+        "[._]*",
+    )
+    """Reserved top-level module names"""
+
+
+def _find_packages_within(root_pkg, pkg_dir):
+    nested = PEP420PackageFinder.find(pkg_dir)
+    return [root_pkg] + [".".join((root_pkg, n)) for n in nested]
+
+
+class ConfigDiscovery:
+    """Fill-in metadata and options that can be automatically derived
+    (from other metadata/options, the file system or conventions)
+    """
+
+    def __init__(self, distribution):
+        self.dist = distribution
+        self._called = False
+        self._root_dir = None  # delay so `src_root` can be set in dist
+
+    def __call__(self, force=False):
+        """Automatically discover missing configuration fields
+        and modifies the given ``distribution`` object in-place.
+
+        Note that by default this will only have an effect the first time the
+        ``ConfigDiscovery`` object is called.
+
+        To repeatedly invoke automatic discovery (e.g. when the project
+        directory changes), please use ``force=True`` (or create a new
+        ``ConfigDiscovery`` instance).
+        """
+        if force is False and self._called:
+            # Avoid overhead of multiple calls
+            return
+
+        self._root_dir = self.dist.src_root or os.curdir
+
+        self._analyse_package_layout()
+        self._analyse_name()  # depends on ``packages`` and ``py_modules``
+
+        self._called = True
+
+    def _analyse_package_layout(self):
+        if self.dist.packages or self.dist.py_modules:
+            # For backward compatibility, just try to find modules/packages
+            # when nothing is given
```

Comment:
This comment strikes me as slightly confusing as to the true intent here, perhaps something like...?
```suggestion
        if self.dist.packages or self.dist.py_modules:
            # For backward compatibility, skip auto-discovery
            # if any values have already been set by caller
```